File: Src/CSharpier.Rider/src/main/java/com/intellij/csharpier/CSharpierProcessPipeMultipleFiles.java
package com.intellij.csharpier;

import com.intellij.openapi.Disposable;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import java.io.*;
import java.nio.charset.Charset;

public class CSharpierProcessPipeMultipleFiles implements ICSharpierProcess, Disposable {

  private final boolean useUtf8;
  private final String csharpierPath;
  private final DotNetProvider dotNetProvider;
  private final String version;
  private Logger logger = CSharpierLogger.getInstance();

  private Process process = null;
  private OutputStreamWriter stdin;
  private BufferedReader stdOut;
  private boolean processFailedToStart;

  public CSharpierProcessPipeMultipleFiles(
    String csharpierPath,
    boolean useUtf8,
    String version,
    Project project
  ) {
    this.csharpierPath = csharpierPath;
    this.useUtf8 = useUtf8;
    this.dotNetProvider = DotNetProvider.getInstance(project);
    this.version = version;
    this.startProcess();

    this.logger.debug("Warm CSharpier with initial format");
    // warm by formatting a file twice, the 3rd time is when it gets really fast
    this.formatFile("public class ClassName { }", "Test.cs");
    this.formatFile("public class ClassName { }", "Test.cs");
  }

  private void startProcess() {
    try {
      var processBuilder = new ProcessBuilder(this.csharpierPath, "--pipe-multiple-files");
      processBuilder.environment().put("DOTNET_NOLOGO", "1");
      processBuilder.environment().put("DOTNET_ROOT", this.dotNetProvider.getDotNetRoot());
      this.process = processBuilder.start();

      var charset = this.useUtf8 ? "utf-8" : Charset.defaultCharset().toString();

      this.stdin = new OutputStreamWriter(this.process.getOutputStream(), charset);
      this.stdOut = new BufferedReader(
        new InputStreamReader(this.process.getInputStream(), charset)
      );

      // if we don't read the error stream, eventually too much is buffered on it and the plugin hangs
      var errorGobbler = new StreamGobbler(this.process.getErrorStream());
      errorGobbler.start();
    } catch (Exception e) {
      this.logger.warn("Failed to spawn the needed csharpier process. Formatting cannot occur.", e);
      this.processFailedToStart = true;
    }
  }

  @Override
  public String getVersion() {
    return this.version;
  }

  @Override
  public boolean getProcessFailedToStart() {
    return this.processFailedToStart;
  }

  @Override
  public String formatFile(String content, String filePath) {
    if (this.processFailedToStart) {
      this.logger.warn("CSharpier proccess failed to start. Formatting cannot occur.");
      return "";
    }

    var stringBuilder = new StringBuilder();

    Runnable task = () -> {
      try {
        this.stdin.write(filePath);
        this.stdin.write('\u0003');
        this.stdin.write(content);
        this.stdin.write('\u0003');
        this.stdin.flush();

        var nextCharacter = this.stdOut.read();
        while (nextCharacter != -1) {
          if (nextCharacter == '\u0003') {
            break;
          }
          stringBuilder.append((char) nextCharacter);
          nextCharacter = this.stdOut.read();
        }
      } catch (Exception e) {
        this.logger.error(e);
        e.printStackTrace();
      }
    };

    // csharpier will freeze in some instances when "Format on Save" is also installed and the file has compilation errors
    // this detects that and recovers from it
    var thread = new Thread(task);
    thread.start();
    try {
      thread.join(3000);
    } catch (InterruptedException e) {
      // if we interrupt it we shouldn't log it
    }

    if (thread.isAlive()) {
      this.logger.warn("CSharpier process appears to be hung, restarting it.");
      thread.interrupt();
      this.process.destroy();
      this.startProcess();
      return "";
    }

    var result = stringBuilder.toString();

    if (result == null || result.isEmpty()) {
      this.logger.info("File is ignored by .csharpierignore or there was an error");
      return "";
    }

    return result;
  }

  @Override
  public void dispose() {
    if (this.process != null) {
      this.process.destroy();
    }
  }

  private class StreamGobbler extends Thread {

    InputStream inputStream;

    private StreamGobbler(InputStream inputStream) {
      this.inputStream = inputStream;
    }

    @Override
    public void run() {
      try {
        var streamReader = new InputStreamReader(this.inputStream);
        var reader = new BufferedReader(streamReader);
        while (reader.readLine() != null) {}
      } catch (IOException ioe) {}
    }
  }
}

File: Src/CSharpier.VisualStudio/CSharpier.VisualStudioShared/CSharpierProcessPipeMultipleFiles.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Text;
using System.Threading.Tasks;
using Process = System.Diagnostics.Process;

namespace CSharpier.VisualStudio
{
    public class CSharpierProcessPipeMultipleFiles : ICSharpierProcess
    {
        private readonly Logger logger;
        private Process process;
        private readonly string csharpierPath;
        private StreamWriter standardIn;

        public string Version { get; }
        public bool ProcessFailedToStart { get; private set; }

        public CSharpierProcessPipeMultipleFiles(
            string csharpierPath,
            string version,
            Logger logger
        )
        {
            this.logger = logger;
            this.csharpierPath = csharpierPath;
            this.Version = version;

            this.StartProcess();

            this.logger.Debug("Warm CSharpier with initial format");
            // warm by formatting a file twice, the 3rd time is when it gets really fast
            this.FormatFile("public class ClassName { }", "Test.cs");
            this.FormatFile("public class ClassName { }", "Test.cs");
        }

        private void StartProcess()
        {
            try
            {
                var processStartInfo = new ProcessStartInfo(
                    this.csharpierPath,
                    " --pipe-multiple-files"
                )
                {
                    RedirectStandardInput = true,
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    StandardOutputEncoding = Encoding.UTF8,
                    StandardErrorEncoding = Encoding.UTF8,
                    UseShellExecute = false,
                    CreateNoWindow = true,
                    EnvironmentVariables = { ["DOTNET_NOLOGO"] = "1" },
                };
                this.process = new Process { StartInfo = processStartInfo };
                this.process.Start();
                this.standardIn = new StreamWriter(
                    this.process.StandardInput.BaseStream,
                    Encoding.UTF8
                );
            }
            catch (Exception ex)
            {
                this.logger.Warn(
                    "Failed to spawn the needed csharpier process. Formatting cannot occur."
                );
                this.logger.Error(ex);
                this.ProcessFailedToStart = true;
            }
        }

        public string FormatFile(string content, string filePath)
        {
            if (this.ProcessFailedToStart)
            {
                this.logger.Warn("CSharpier process failed to start. Formatting cannot occur.");
                return "";
            }

            var stringBuilder = new StringBuilder();

            var task = Task.Run(() =>
            {
                this.standardIn.Write(filePath);
                this.standardIn.Write('\u0003');
                this.standardIn.Write(content);
                this.standardIn.Write('\u0003');
                this.standardIn.Flush();

                var nextCharacter = this.process.StandardOutput.Read();
                while (nextCharacter != -1)
                {
                    if (nextCharacter == '\u0003')
                    {
                        break;
                    }

                    stringBuilder.Append((char)nextCharacter);
                    nextCharacter = this.process.StandardOutput.Read();
                }
            });

            // csharpier will freeze in some instances when "Format on Save" is also installed and the file has compilation errors
            // this detects that and recovers from it
            if (!task.Wait(TimeSpan.FromSeconds(3)))
            {
                this.logger.Warn("CSharpier process appears to be hung, restarting it.");
                this.process.Kill();
                this.StartProcess();
                return string.Empty;
            }

            var result = stringBuilder.ToString();

            if (string.IsNullOrEmpty(result))
            {
                this.logger.Debug("File is ignored by .csharpierignore or there was an error");
            }

            return result;
        }

        public void Dispose()
        {
            this.process.Kill();
        }
    }
}
